internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,433 @@
+				FailedAttempts: 0,
+			})
+		}
+	}
+
+	manager.DynamicClientWrappers = newDynamicClients
+	manager.lastFallbackClientRefresh = metav1.Now()
+	return nil
+}
+
+func (manager *ConfigurationClientManager) UpdateClientBackoffStatus(clientWrapper *ConfigurationClientWrapper, successful bool) {
+	if successful {
+		clientWrapper.BackOffEndTime = metav1.Time{}
+		clientWrapper.FailedAttempts = 0
+	} else {
+		clientWrapper.FailedAttempts++
+		clientWrapper.BackOffEndTime = metav1.Time{Time: metav1.Now().Add(calculateBackoffDuration(clientWrapper.FailedAttempts))}
+	}
+}
+
+func QuerySrvTargetHost(host string) ([]string, error) {
+	results := make([]string, 0)
+
+	_, originRecords, err := net.LookupSRV(Origin, TCP, host)
+	if err != nil {
+		// If the host does not have SRV records => no replicas
+		return results, err
+	}
+
+	originHost := strings.TrimSuffix(originRecords[0].Target, ".")
+	index := 0
+	for {
+		currentAlt := Alt + strconv.Itoa(index)
+		_, altRecords, err := net.LookupSRV(currentAlt, TCP, originHost)
+		if err != nil {
+			// If the host does not have SRV records => no more replicas
+			if _, ok := err.(*net.DNSError); ok {
+				break
+			} else {
+				return nil, err
+			}
+		}
+
+		for _, record := range altRecords {
+			altHost := strings.TrimSuffix(record.Target, ".")
+			if altHost != "" {
+				results = append(results, altHost)
+			}
+		}
+		index = index + 1
+	}
+
+	return results, nil
+}
+
+func (manager *ConfigurationClientManager) newConfigurationClient(endpoint string) (*azappconfig.Client, error) {
+	if manager.credential != nil {
+		return azappconfig.NewClient(endpoint, manager.credential, clientOptionWithModuleInfo)
+	}
+
+	connectionStr := buildConnectionString(endpoint, manager.secret, manager.id)
+	if connectionStr == "" {
+		return nil, fmt.Errorf("failed to build connection string for fallback client")
+	}
+
+	return azappconfig.NewClientFromConnectionString(connectionStr, clientOptionWithModuleInfo)
+}
+
+func isValidEndpoint(host string, validDomain string) bool {
+	if validDomain == "" {
+		return false
+	}
+
+	return strings.HasSuffix(strings.ToLower(host), strings.ToLower(validDomain))
+}
+
+func getValidDomain(endpoint string) string {
+	url, _ := url.Parse(endpoint)
+	TrustedDomainLabels := []string{AzConfigDomainLabel, AppConfigDomainLabel}
+	for _, label := range TrustedDomainLabels {
+		index := strings.LastIndex(strings.ToLower(url.Host), strings.ToLower(label))
+		if index != -1 {
+			return url.Host[index:]
+		}
+	}
+
+	return ""
+}
+
+func buildConnectionString(endpoint string, secret string, id string) string {
+	if secret == "" || id == "" {
+		return ""
+	}
+
+	return fmt.Sprintf("%s=%s;%s=%s;%s=%s",
+		EndpointSection, endpoint,
+		IdSection, id,
+		SecretSection, secret)
+}
+
+func parseConnectionString(connectionString string, token string) (string, error) {
+	if connectionString == "" {
+		return "", NewArgumentError("connectionString", fmt.Errorf("connectionString cannot be empty"))
```

Comment:
Should not return as ArugmentError in this method, should handle all the errors as runtime error.